lib/std/zig/render.zig

```diff
@@ -1710,8 +1728,31 @@ fn renderFnProto(r: *Render, fn_proto: Ast.full.FnProto, space: Space) Error!voi
 
     // The params list is a sparse set that does *not* include anytype or ... parameters.
 
-    const trailing_comma = token_tags[rparen - 1] == .comma;
-    if (!trailing_comma and !hasComment(tree, lparen, rparen)) {
+    const src_has_trailing_comma = token_tags[rparen - 1] == .comma;
+    const one_line = if (src_has_trailing_comma) false else one_line: {
+        // We print all the members in-line unless one of the following conditions are true:
+
+        // 1. The prototype has comments.
+        if (hasComment(tree, lparen, rparen)) {
+            break :one_line false;
+        }
+
+        // 2. A parameter of the prototype has a doc comment.
+        var last_param_token = lparen;
+        while (true) {
+            last_param_token += 1;
+            switch (token_tags[last_param_token]) {
+                .doc_comment => break :one_line false,
+                .ellipsis3 => break,
+                .r_paren => break,
+                else => {},
+            }
+        }
+
+        break :one_line true;
+    };
```

Comment:
Good find. This undoes the `testEpoch` formatting.

Reply:
I couldn't come up with a case that breaks when not using `tree.lastToken` and considering https://github.com/ziglang/zig/pull/20116#issuecomment-2139687641 I don't see a problem here so I'm marking this as resolved for now.

Reply:
Comments in an inner scope affecting the formatting of the outer scope seems like a bug. I don't think introducing an identical bug for consistency is desirable.

Reply:
Not sure if it is a bug. Please open a proposal or a bug report and get it assessed by someone that knows.

Reply:
@wooster0 The current implementation is still wrong and broken. For example, the following code which contains a closing parenthesis
```zig
fn foo(
    bar: (i32),
    /// The second operand
    baz: i32
) i32 {
    return bar + baz;
}
```
will get mangled into
```zig
fn foo(bar: (i32), i32) i32 {
    return bar + baz;
}
```
(possibly because the `.doc_comment => unreachable` prong is reached).

You need to process tokens correctly using `tree.lastToken(param)` for the correct result.

Please unresolve this conversation until this has been fixed.

Reply:
Thanks for the test case.

Reply:
Fixed